File: tree/98.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        
        #inorder traversal
        
        #time: O(n)
        #space: O(1)
        
        if root == None:
            return True
        
        previous = - math.inf
        valid = True
        
        def recursive(node):
        
        
            if node:
                recursive(node.left)
                
                if node.val <= previous:
                    valid = False
                previous = node.val
                
                recursive(node.right)
                
       
        recursive(root)
        return valid

File: grind75/Dec29/98.go
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool helper(TreeNode* root, long minval, long maxval) {
        if (!root) { return true; }
        if (root->val <= minval || root->val >= maxval) {
            return false;
        }
        return helper(root->left, minval, root->val) && helper(root->right, root->val, maxval);
    }


    bool isValidBST(TreeNode* root) {
        if (!root) { return true; }
        return helper(root->left, LONG_MIN, root->val) && helper(root->right, root->val, LONG_MAX);
    }
};